### ISSUE TITLE
mega: option to use `mega-cmd` internally to access storage

### ISSUE THREAD

**Issue by msalini** [rclone/rclone#3660]
#### What problem are you are trying to solve?
Given the ongoing issues with Mega connections (as reported in #3393) , rclone can delegate the Mega authentication to [mega-cmd](https://mega.nz/cmd) for those users who have mega-cmd installed and connect to Mega via webdav through the webdav service initiated by mega-cmd.

#### How do you think rclone should be changed to solve that?
When a new Mega instance is created in rclone:

1. rclone checks whether mega-cmd is installed
2. If so, rclone passes the authentication details to mega-cmd and starts up a mega-cmd webdav demon (mega-cmd webdav guide [here](https://github.com/meganz/MEGAcmd/blob/master/contrib/docs/WEBDAV.md))
3. rclone then initiates a webdav connection to the demon via localhost (instructions on how to do this manually are [here](https://github.com/rclone/rclone/issues/3393#issuecomment-545119027))
4. If mega-cmd is not installed, rclone can initiate a traditional mega connection with a warning/recommendation to install mega-cmd.

**Comment by ncw** [comment]
This is a great idea thanks, and webdav through mega-cmd works really quite well.

I can see having a webdav mode which runs mega-cmd for you would be relatively easy.

**Comment by caspertone2003** [comment]
Please do not deprecate current mega mode... yet... Mounts in windows have their own kind of issues...

**Comment by caspertone2003** [comment]
Caveat
As I expected, current megaCMD is not able to work with two mega remotes at same time...
This means that such functionality needs to be incorporated or wait to be developed by mega...
Someone already asked recently 
https://github.com/meganz/MEGAcmd/issues/264
CT

**Comment by kidburglar** [comment]
Hello, I saw this issue and tested the webdav and ftp because I started to experience the same issue but I ot think it work really well for "mass uploads".

For what I saw it make a temp fiie on the webdav/ftp from mega-cmd and it upload afterword 1 by 1.
It mean that between rclone (process) and megacmd (upload) there will still be a wait time for rclone.

Does you experience it too with your test or maybe I missed a configuration somewhere (or missunderstand something)?

**Comment by ivandeex** [comment]
A similar idea for _cryptomator_: https://github.com/rclone/rclone/issues/1647#issuecomment-735499029

**Comment by ivandeex** [comment]
https://github.com/rclone/rclone/issues/3393#issuecomment-545352586

> What I could do is make a "mega" mode for the webdav backend which can invoke mega-cmd automatically.
> I don't think I'd remove the mega backend, maybe just deprecate it.

What about making this a new backend rather than a webdav extension?

**Comment by ncw** [comment]
> [#3393 (comment)](https://github.com/rclone/rclone/issues/3393#issuecomment-545352586)
> 
> > What I could do is make a "mega" mode for the webdav backend which can invoke mega-cmd automatically.
> > I don't think I'd remove the mega backend, maybe just deprecate it.
> 
> What about making this a new backend rather than a webdav extension?

We could wrap the webdav backend, like the hubic backend wraps the swift backen.
 
> If we ever release it, we can recommend it and deprecate the old backend.

:-)

**Comment by CodingKoopa** [comment]
I have documented some of the issues that occurred when I tried using MEGAcmd's WebDAV implementation with rclone [here](https://github.com/rclone/rclone/issues/4695#issuecomment-1006262360) (in #4695).

**Comment by kadogo** [comment]
Hello @ncw 

It is a little old, but I'm reading about the issues related to mega when I found this one.

It would not be a good use about an "executable" backend that I talked about in MP (on forum.rclone.org).
I still have that i my mind, but I'm not good enough in Go for doing it myself.

If you think it's a good idea maybe I (or someone else) can make an issue about it and hope for a PR?

**Comment by ImportTaste** [comment]
@ncw I'm not positive this will work as I'm not overly familiar with sessions, but since MEGAcmd can retrieve your active session ID, why not allow the session ID to be part of the rclone command line so that rclone can use the same one?

**Comment by ncw** [comment]
@ImportTaste I don't think this is possible with the way the mega library is laid out at the moment. What rclone should be doing is opening a session and keeping the session on disk along with all the files. This would make rclone a lot quicker but it would be a lot of work in the mega lib too :-(
